tests/unit/s2n_mem_testlib_test.c

```diff
@@ -0,0 +1,255 @@
+        EXPECT_NOT_NULL(mem.data);
+        EXPECT_EQUAL(mem.size, requested);
+
+        /* Test: s2n_mem_test_get_malloc_count */
+        {
+            uint32_t count = 0;
+            EXPECT_OK(s2n_mem_test_get_malloc_count(&count));
+            EXPECT_EQUAL(count, 1);
+        };
+
+        /* Test: s2n_mem_test_get_malloc */
+        {
+            struct s2n_mem_test_malloc malloc = { 0 };
+            EXPECT_OK(s2n_mem_test_get_malloc(0, &malloc));
+            EXPECT_EQUAL(malloc.requested, requested);
+            EXPECT_FALSE(malloc.freed);
+
+            EXPECT_ERROR_WITH_ERRNO(s2n_mem_test_get_malloc(1, &malloc),
+                    S2N_ERR_STUFFER_OUT_OF_DATA);
+        };
+
+        /* Test: s2n_mem_test_assert_malloc_count */
+        {
+            EXPECT_OK(s2n_mem_test_assert_malloc_count(1));
+            EXPECT_ERROR_WITH_ERRNO(s2n_mem_test_assert_malloc_count(0),
+                    S2N_ERR_TEST_ASSERTION);
+            EXPECT_ERROR_WITH_ERRNO(s2n_mem_test_assert_malloc_count(2),
+                    S2N_ERR_TEST_ASSERTION);
+        };
+
+        /* Test: s2n_mem_test_assert_malloc */
+        {
+            EXPECT_OK(s2n_mem_test_assert_malloc(requested));
+            EXPECT_ERROR_WITH_ERRNO(s2n_mem_test_assert_malloc(0), S2N_ERR_TEST_ASSERTION);
+            EXPECT_ERROR_WITH_ERRNO(s2n_mem_test_assert_malloc(1), S2N_ERR_TEST_ASSERTION);
+            EXPECT_ERROR_WITH_ERRNO(s2n_mem_test_assert_malloc(UINT32_MAX),
+                    S2N_ERR_TEST_ASSERTION);
+        };
+
+        /* Test: s2n_mem_test_assert_all_freed */
+        {
+            EXPECT_ERROR_WITH_ERRNO(s2n_mem_test_assert_all_freed(), S2N_ERR_TEST_ASSERTION);
+            EXPECT_SUCCESS(s2n_free(&mem));
+            EXPECT_OK(s2n_mem_test_assert_all_freed());
+        };
+    };
+
+    /* Test: Multiple mallocs */
+    {
+        DEFER_CLEANUP(struct s2n_mem_test_cb_scope scope = { 0 },
+                s2n_mem_test_free_callbacks);
+        EXPECT_OK(s2n_mem_test_init_callbacks(&scope));
+
+        const size_t requested[] = { 10, 1, 17, 17, 111 };
+        const size_t count = s2n_array_len(requested);
+        for (size_t i = 0; i < count; i++) {
+            DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+            EXPECT_SUCCESS(s2n_alloc(&mem, requested[i]));
+            EXPECT_NOT_NULL(mem.data);
+            EXPECT_EQUAL(mem.size, requested[i]);
+        }
+
+        /* Test: s2n_mem_test_get_malloc_count */
+        {
+            uint32_t actual_count = 0;
+            EXPECT_OK(s2n_mem_test_get_malloc_count(&actual_count));
+            EXPECT_EQUAL(count, actual_count);
+        };
+
+        /* Test: s2n_mem_test_get_malloc */
+        {
+            struct s2n_mem_test_malloc malloc = { 0 };
+            for (size_t i = 0; i < count; i++) {
+                EXPECT_OK(s2n_mem_test_get_malloc(i, &malloc));
+                EXPECT_EQUAL(malloc.requested, requested[i]);
+                EXPECT_TRUE(malloc.freed);
```

Comment:
I alloced each blob inside the for loop, so they got cleaned up inside the for loop too. Otherwise I'd have to keep around an array of blobs to free later.

Reply:
Would it be clearer if I wrote the loop like:
```
        struct s2n_blob mem = { 0 };
        for (size_t i = 0; i < count; i++) {
            EXPECT_SUCCESS(s2n_alloc(&mem, requested[i]));
            EXPECT_NOT_NULL(mem.data);
            EXPECT_EQUAL(mem.size, requested[i]);
            EXPECT_SUCCESS(s2n_free(&mem));
        }
```

Reply:
Yeah, it would be more clear. It can be hard to keep track of all of the affects of the defer_cleanup.

Reply:
Yeah, that's fair. Probably best to make the free explicit if I'm testing for it.